KABOT-Project/Scratch-Project
# Dk/authentication

Overview

**Issue Type**
-[] Bug
-[x] Feature
-[]Tech debt

**Description**
-Created a startSession middleware that starts the session (just to modularize code); everything else is the same, except that req.session.user_id is assigned within this middleware (removed from other places). 

-Created verifyUser middleware (checks if authentication = true on the sessions object as a way to authenticate sessions)
 
-Added a logout button on the homepage that closes the session and redirects to homepage (can definitely be restyled/moved)

-A ‘provider’ column has been added to the users table to handle different types of authentication. The code should be such that if users sign up through the app, provider = “app”; if signed up through Google OAuth, provider = “GoogleOAuth”. 

-For Google OAuth: when users who already have an account made through OAuth sign in, they should be directed to the homepage. For users who don’t have an account, a new entry in the database is created for them.  

**Steps to Reproduce Bug/Validate Feature/Confirm Tech Debt Fix**
Go through Google OAuth steps and check console to see if the queries are working as intended. 

**Previous behavior**
n/a

**Expected behavior**
Google OAuth should create users/login users. 
If someone isn't logged in, they should not have access to other pages and should be redirected to /welcome. 

**Screenshots & Videos**
n/a

**Additional context**
In the users table, I removed the NOT NULL constraint from the password column, to allow Google OAuth to leave that blank when signing up for a user. One of my next goals is to implement some basic functionality to require the password on the front-end in order to submit.
